### PR TITLE
Implement timestamp vote triggering.

### PR DESCRIPTION
**Pull request by piotrm50**
# Description of change

Trigger FPC vote on message timestamp when LoK = 1

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/consensus/fcob/storage.go [hunk 1]
```diff
@@ -580,6 +583,22 @@ func (t *TimestampOpinion) ObjectStorageValue() []byte {
 		Bytes()
 }
 
+// SetLiked sets the opinion's liked.
```
Comment (apenzk) [inline packages/consensus/fcob/storage.go:586]:
```suggestion
// SetLiked sets the opinion's like status.
```

File: packages/consensus/fcob/storage.go [hunk 2]
```diff
@@ -580,6 +583,22 @@ func (t *TimestampOpinion) ObjectStorageValue() []byte {
...
+func (t *TimestampOpinion) SetLiked(newOpinion opinion.Opinion) {
```
Comment (apenzk) [inline packages/consensus/fcob/storage.go:587]: This should not go into FCoB because it is not FCoB related

File: packages/vote/vote_context.go [hunk 3]
```diff
@@ -16,7 +16,7 @@ type ObjectType uint8
 
 const (
 	// ConflictType defines an object type conflict.
-	ConflictType = iota
+	ConflictType ObjectType = iota
```
Comment (apenzk) [inline packages/vote/vote_context.go:19]: why is this needed now - did it not work before?

File: plugins/analysis/dashboard/fpc_storage.go [hunk 4]
```diff
@@ -25,6 +27,8 @@ type FPCRecord struct {
 	Opinions []int32 `json:"opinions" bson:"opinions"`
 	// Outcome defines final opinion of the conflict.
 	Outcome int32 `json:"outcome" bson:"outcome"`
+	// Type defines the type of conflict - timestamp or payload.
```
Comment (apenzk) [inline plugins/analysis/dashboard/fpc_storage.go:30]:
```suggestion
	// Type defines the type of vote - timestamp or conflict.
```

File: plugins/messagelayer/parameters.go [hunk 5]
```diff
@@ -20,7 +20,9 @@ var Parameters = struct {
 
 	// FCOB contains parameters related to the fast consensus of barcelona.
 	FCOB struct {
-		AverageNetworkDelay int `default:"5" usage:"the avg. network delay to use for FCoB rules"`
+		AverageNetworkDelay    int `default:"5" usage:"the avg. network delay to use for FCoB rules in seconds"`
+		TimestampWindow        int `default:"60" usage:"the time window for assessing the timestamp quality in seconds"`
```
Comment (apenzk) [inline plugins/messagelayer/parameters.go:24]:
```suggestion
		TimestampWindow        int `default:"60" usage:"the time window for setting the initial timestamp opinion as liked in seconds"`
```

File: packages/consensus/fcob/consensusmechanism.go [hunk 6]
```diff
@@ -99,13 +100,28 @@ func (f *ConsensusMechanism) Evaluate(messageID tangle.MessageID) {
...
+
+	if !f.tangle.Storage.Message(messageID).Consume(func(messageMetadata *tangle.Message) {
+		issuingTime = messageMetadata.IssuingTime()
+	}) {
+		// FIXME: what to do when message does not exist in the storage?
+		return
+	}
+
+	timestampOpinion, _ := TimestampQuality(messageID, issuingTime, clock.SyncedTime())
+	// TODO: handle error which indicates timestamp from the future
```
[the remark this answers is not in the document]
Comment (piotrm50) [reply packages/consensus/fcob/consensusmechanism.go:116]: If the panic is implemented in TimestampQuality then I think it's fine, but then is error ever returned?

File: packages/vote/vote_context.go (hunk 3 above)
Comment (piotrm50) [reply packages/vote/vote_context.go:19]: Before those values were not used when triggering the FPC vote as it was hardcoded to always use ConflictType. Now that we need to parametrize that there were problems with the typing.

File: packages/consensus/fcob/consensusmechanism.go [hunk 7]
```diff
@@ -133,13 +150,42 @@ func (f *ConsensusMechanism) ProcessVote(ev *vote.OpinionEvent) {
...
 			}
 		})
+	case vote.TimestampType:
+		messageID, err := tangle.NewMessageID(ev.ID)
+		if err != nil {
+			f.Events.Error.Trigger(err)
+			return
+		}
+
+		f.Storage.TimestampOpinion(messageID).Consume(func(opinion *TimestampOpinion) {
```
[the remark this answers is not in the document]
Comment (piotrm50) [reply packages/consensus/fcob/consensusmechanism.go:160]: Done

File: packages/consensus/fcob/consensusmechanism.go [hunk 8]
```diff
@@ -133,13 +150,42 @@ func (f *ConsensusMechanism) ProcessVote(ev *vote.OpinionEvent) {
...
+			opinion.SetLevelOfKnowledge(Two)
+		})
+
+		// eligible when timestamp is ok but confirmed when timestamp and tx is correct
+		f.setEligibility(messageID)
+
+		f.setTimestampOpinionDone(messageID)
+
+		if f.messageDone(messageID) {
+			f.tangle.Utils.WalkMessageID(f.createMessageOpinion, tangle.MessageIDs{messageID}, true)
```
[the remark this answers is not in the document]
Comment (piotrm50) [reply packages/consensus/fcob/consensusmechanism.go:171]: Done

File: packages/consensus/fcob/consensusmechanism.go [hunk 9]
```diff
@@ -133,13 +150,42 @@ func (f *ConsensusMechanism) ProcessVote(ev *vote.OpinionEvent) {
...
+			opinion.SetLiked(ev.Opinion)
+			opinion.SetLevelOfKnowledge(Two)
+		})
+
+		// eligible when timestamp is ok but confirmed when timestamp and tx is correct
+		f.setEligibility(messageID)
+
+		f.setTimestampOpinionDone(messageID)
+
+		if f.messageDone(messageID) {
```
[the remark this answers is not in the document]
Comment (piotrm50) [reply packages/consensus/fcob/consensusmechanism.go:170]: Then this part will not be executed. It will be executed instead when e.g. voting about a payload finishes. Or in case of a correct message, it will be marked as confirmed either in timestamp or payload evaluation, depending on which one will be executed second.

File: packages/consensus/fcob/storage.go (hunk 2 above)
Comment (piotrm50) [reply packages/consensus/fcob/storage.go:587]: I have it all refactored on a separate branch, to make transaction and message opinions be similar in terms of implemention.

File: packages/consensus/fcob/consensusmechanism.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (apenzk) [reply packages/consensus/fcob/consensusmechanism.go:170]: ok thanks. I think this would be good information to have here. could you add this.

File: packages/consensus/fcob/consensusmechanism.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (apenzk) [reply packages/consensus/fcob/consensusmechanism.go:171]: i cant seem to find it

File: packages/consensus/fcob/storage.go (hunk 2 above)
Comment (apenzk) [reply packages/consensus/fcob/storage.go:587]: ok

File: packages/consensus/fcob/consensusmechanism.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (apenzk) [reply packages/consensus/fcob/consensusmechanism.go:116]: so we need to log it and then panic? i think the best place for putting the panic is in timestampQuality, since we already  create "diff" there

[the remark this answers is not in the document]
Comment (apenzk) [reply packages/consensus/fcob/consensusmechanism.go:116]: discussion ended with concluding to setting it to like

File: packages/consensus/fcob/consensusmechanism.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (piotrm50) [reply packages/consensus/fcob/consensusmechanism.go:170]: When I was addressing your comments I created another branch with refactoring of lots of the code and it's there. I will create another PR with those changes once this one gets merged not to cause confusion regarding the actual changes here.
